File: migrator/orchestrator.py
import logging
import signal
import threading
import time
from collections import defaultdict
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timezone
from typing import Callable, Optional

import psycopg2
import psycopg2.extras

from migrator.cast import try_cast_string
from migrator.config import MigratorConfig
from migrator.partition import Partitioning, compute_partition
from migrator.pg_reader import PgReader
from migrator.progress import ProgressTracker
from migrator.scylla_writer import ScyllaWriter

logger = logging.getLogger(__name__)


class Orchestrator:
    def __init__(
        self,
        config: MigratorConfig,
        reader: PgReader,
        writer: ScyllaWriter,
        tracker: ProgressTracker,
        conn_factory: Optional[Callable] = None,
    ):
        self._cfg = config
        self._reader = reader
        self._writer = writer
        self._tracker = tracker
        self._conn_factory = conn_factory
        self._stop = False
        self._partitioning = Partitioning(config.partitioning)
        self._partition_fn = lambda ts: compute_partition(ts, self._partitioning)
        self._cast_fn = try_cast_string if config.cast_strings else None

    def run(self, historical_only: bool = False, resume: bool = False):
        """Entry point: Phase 0 → Phase 1 → Phase 2 (unless historical_only)."""
        signal.signal(signal.SIGTERM, self._handle_sigterm)
        signal.signal(signal.SIGINT, self._handle_sigterm)

        if resume and self._tracker.load():
            logger.info("Resuming from checkpoint: phase=%s", self._tracker.progress.phase)
        else:
            self._tracker.update(
                phase="phase0",
                partitioning=self._cfg.partitioning,
                cast_strings=self._cfg.cast_strings,
                started_at=datetime.now(timezone.utc).isoformat(),
            )

        entity_map, key_map = self._phase0()
        self._phase1(entity_map, key_map)

        if not historical_only and not self._stop:
            self._phase2(entity_map, key_map)

    def _handle_sigterm(self, signum, frame):
        logger.info("Signal %s received — stopping after current batch", signum)
        self._stop = True

    def _phase0(self):
        """Load entity map and key map. Returns (entity_map, key_map)."""
        logger.info("Phase 0: loading entity map and key map")
        entity_map = self._reader.load_entity_map()
        key_map = self._reader.load_key_map()
        total = self._reader.count_ts_kv()
        logger.info(
            "Phase 0 complete: %d entities, %d keys, %d ts_kv rows",
            len(entity_map),
            len(key_map),
            total,
        )
        now_ms = int(time.time() * 1000)
        self._tracker.update(phase="phase1", phase1_start_ts=now_ms)
        return entity_map, key_map

    def _phase1(self, entity_map, key_map):
        """Historical migration: all ts_kv rows → ScyllaDB, using parallel workers."""
        workers = self._cfg.workers if self._conn_factory else 1
        logger.info("Phase 1: historical migration started (workers=%d)", workers)

        progress = self._tracker.progress
        completed = set(progress.completed_entities)

        all_entities = list(self._reader.iter_distinct_entities())
        pending = [e for e in all_entities if e not in completed]
        logger.info(
            "Phase 1: %d entities total, %d pending, %d already done",
            len(all_entities), len(pending), len(completed),
        )

        lock = threading.Lock()

        def process_entity(entity_id_str: str):
            if self._stop:
                return

            entity_type = entity_map.get(entity_id_str)
            if entity_type is None:
                logger.warning("entity_id %s not found in entity_map, skipping", entity_id_str)
                with lock:
                    progress.skipped_rows += 1
                    self._tracker.update(skipped_rows=progress.skipped_rows)
                return

            conn = self._conn_factory()
            try:
                reader = PgReader(conn)
                batch_rows = []
                for row in reader.iter_ts_kv_for_entity(entity_id_str, self._cfg.batch_size):
                    if self._stop:
                        break
                    batch_rows.append(row)
                    if len(batch_rows) >= self._cfg.batch_size:
                        self._flush_ts_batch(batch_rows, entity_type, key_map)
                        with lock:
                            progress.migrated_rows += len(batch_rows)
                            self._tracker.update(migrated_rows=progress.migrated_rows)
                        batch_rows = []

                if batch_rows and not self._stop:
                    self._flush_ts_batch(batch_rows, entity_type, key_map)
                    with lock:
                        progress.migrated_rows += len(batch_rows)

                if not self._stop:
                    with lock:
                        progress.completed_entities.append(entity_id_str)
                        self._tracker.update(
                            migrated_rows=progress.migrated_rows,
                            last_entity_id=entity_id_str,
                            completed_entities=progress.completed_entities,
                        )
            finally:
                conn.close()

        if workers > 1:
            with ThreadPoolExecutor(max_workers=workers) as executor:
                futures = {executor.submit(process_entity, eid): eid for eid in pending}
                for future in as_completed(futures):
                    if self._stop:
                        for f in futures:
                            f.cancel()
                        break
                    exc = future.exception()
                    if exc:
                        logger.error("Entity %s failed: %s", futures[future], exc, exc_info=exc)
        else:
            for eid in pending:
                if self._stop:
                    break
                process_entity(eid)

        # Migrate latest values (single-threaded sequential scan)
        logger.info("Phase 1: migrating ts_kv_latest")
        latest_by_type = defaultdict(list)
        for row in self._reader.iter_ts_kv_latest():
            entity_id_str = str(row["entity_id"])
            entity_type = entity_map.get(entity_id_str)
            if entity_type is None:
                continue
            latest_by_type[entity_type].append(row)
            if len(latest_by_type[entity_type]) >= self._cfg.batch_size:
                self._writer.write_latest_batch(
                    latest_by_type[entity_type], entity_type, key_map, self._cast_fn
                )
                latest_by_type[entity_type] = []

        for entity_type, rows in latest_by_type.items():
            if rows:
                self._writer.write_latest_batch(rows, entity_type, key_map, self._cast_fn)

        logger.info("Phase 1 complete")
        self._tracker.update(phase="live_sync")

    def _flush_ts_batch(self, rows, entity_type, key_map):
        """Write a batch and its partitions to ScyllaDB. Thread-safe."""
        partitions = self._writer.write_ts_batch(
            rows, entity_type, key_map, self._partition_fn, self._cast_fn
        )
        self._writer.write_partitions(partitions)

    def _phase2(self, entity_map, key_map):
        """Live sync: poll ts_kv for rows newer than watermark."""
        logger.info("Phase 2: live sync started")
        progress = self._tracker.progress
        watermark = progress.watermark_ts or (progress.phase1_start_ts - 60_000)
        self._tracker.update(watermark_ts=watermark)

        while not self._stop:
            now_ms = int(time.time() * 1000)
            batch_rows_by_type = defaultdict(list)
            latest_by_type = defaultdict(list)

            for row in self._reader.iter_ts_kv_by_ts(watermark, self._cfg.batch_size):
                entity_id_str = str(row["entity_id"])
                entity_type = entity_map.get(entity_id_str)
                if entity_type is None:
                    continue
                batch_rows_by_type[entity_type].append(row)
                latest_by_type[entity_type].append(row)
                if row["ts"] > watermark:
                    watermark = row["ts"]

            for entity_type, rows in batch_rows_by_type.items():
                self._flush_ts_batch(rows, entity_type, key_map)

            for entity_type, rows in latest_by_type.items():
                self._writer.write_latest_batch(rows, entity_type, key_map, self._cast_fn)

            lag = now_ms - watermark
            self._tracker.update(watermark_ts=watermark)

            if lag < self._cfg.lag_threshold_ms:
                logger.info("LAG %dms < threshold %dms — ready for switchover", lag, self._cfg.lag_threshold_ms)

            time.sleep(self._cfg.live_sync_interval)

File: Orchestrator.cs
using Npgsql;

namespace TbMigrator;

public class Orchestrator
{
    private readonly AppConfig       _cfg;
    private readonly PgReader        _reader;
    private readonly ScyllaWriter    _scylla;
    private readonly ProgressTracker _tracker;
    private readonly string          _connStr;
    private volatile bool            _stop;

    public Orchestrator(AppConfig cfg, PgReader reader, ScyllaWriter scylla, ProgressTracker tracker)
    {
        _cfg     = cfg;
        _reader  = reader;
        _scylla  = scylla;
        _tracker = tracker;
        _connStr = cfg.Pg.ConnectionString;
    }

    public async Task RunAsync(bool historicalOnly, bool resume, CancellationToken ct)
    {
        if (resume && _tracker.Load())
            Console.Error.WriteLine($"[INFO] Resuming: phase={_tracker.Progress.Phase}");
        else
            _tracker.Update(p =>
            {
                p.Phase        = "phase0";
                p.Partitioning = _cfg.Migrator.Partitioning;
                p.CastStrings  = _cfg.Migrator.CastStrings;
                p.StartedAt    = DateTimeOffset.UtcNow.ToString("O");
            });

        var (entityMap, keyMap, hybridMode) = await Phase0Async(ct);
        await Phase1Async(entityMap, keyMap, hybridMode, ct);

        if (!historicalOnly && !_stop)
            await Phase2Async(entityMap, keyMap, hybridMode, ct);
    }

    // -------------------------------------------------------------------------
    // Partition runner — migrate one validated partition (entity-type resolved
    // via existing entity map). Stream batches sequentially (ordering matters
    // for keyset pagination), write batches in parallel via ScyllaWriter
    // semaphore + Task.WhenAll.
    // NOTE: `workers` is intentionally unused — streaming stays sequential by design
    // (keyset pagination ordering); parallelism comes from Task.WhenAll + ScyllaWriter
    // semaphore. Kept for CLI signature consistency and future use.
    // -------------------------------------------------------------------------
    public async Task RunPartitionAsync(string partition, long deltaFromTs, int workers, CancellationToken ct)
    {
        var (keyMap, hybridMode) = await _reader.LoadKeyMapAsync(ct);
        var entityMap = await _reader.LoadEntityMapAsync(ct);
        long pgCount = await _reader.CountPartitionAsync(partition, ct);

        // Modes:
        //   DELTA (--delta-from) — ts-ordered stream of rows with ts > filter.
        //     Used for hot-partition catch-up while TB is writing. Counter adds
        //     to the previous run (rows before the filter are already in Scylla).
        //   FULL (no --delta-from) — PK-ordered stream (entity_id, key, ts):
        //     the PK index serves the scan directly — O(N). Scylla INSERTs are
        //     idempotent, so a fresh full pass self-corrects the counter.
        //     Resume: checkpoint cursor (LastEntityId/LastKey/LastTs) continues
        //     exactly where an interrupted FULL pass stopped.
        bool deltaMode = deltaFromTs != long.MinValue;
        long baseCount = 0, prevMaxTs = 0;
        string? prevDump = null;
        (Guid EntityId, string KeyRaw, long Ts)? resumeCursor = null;
        if (_tracker.Progress.Partitions.TryGetValue(partition, out var prev))
        {
            prevDump = prev.DumpFile;
            prevMaxTs = prev.MaxTs;
            if (deltaMode)
            {
                baseCount = prev.ScyllaCount;
            }
            else if (prev.State == "migrating"
                     && prev.LastEntityId is { Length: > 0 } leid
                     && prev.LastKey is { Length: > 0 } lkey
                     && Guid.TryParse(leid, out var leidGuid))
            {
                baseCount = prev.ScyllaCount;
                resumeCursor = (leidGuid, lkey, prev.LastTs);
            }
            // else: legacy checkpoint without cursor or completed pass → full
            // idempotent pass from zero (self-correcting counter).
        }
        _tracker.Update(p =>
        {
            p.Phase = $"partition:{partition}";
            p.Partitions[partition] = new PartitionProgress(
                "migrating", pgCount, baseCount, prevDump, false, false, prevMaxTs);
        });

        // Reverse key map: TsRow.Key holds the resolved key NAME; the resume
        // cursor needs the RAW key (physical column value). Key names are unique
        // in ts_kv_dictionary, so the reverse lookup is exact.
        Dictionary<string, string>? keyNameToRaw = hybridMode
            ? keyMap.ToDictionary(kv => kv.Value, kv => kv.Key.ToString())
            : null;

        long written = 0, maxTs = prevMaxTs;

        await foreach (var batch in ReadBatchesAsync(partition, deltaFromTs, resumeCursor, keyMap, hybridMode, ct))
        {
            var byType = new Dictionary<string, List<TsRow>>();
            foreach (var row in batch)
            {
                if (!entityMap.TryGetValue(row.EntityId, out var et)) continue;
                if (!byType.ContainsKey(et)) byType[et] = [];
                byType[et].Add(row);
                if (row.Ts > maxTs) maxTs = row.Ts;
            }
            var tasks = byType.Select(async kv =>
            {
                var parts = await _scylla.WriteTsBatchAsync(kv.Value, kv.Key, _cfg.Migrator.Partitioning, _cfg.Migrator.CastStrings, ct);
                await _scylla.WritePartitionsAsync(parts, ct);
            });
            await Task.WhenAll(tasks);
            written += batch.Count;
            var lastRow = batch[^1];
            string lastRawKey = keyNameToRaw is null
                ? lastRow.Key
                : keyNameToRaw.TryGetValue(lastRow.Key, out var raw) ? raw : lastRow.Key;
            long w = baseCount + written, m = maxTs;
            _tracker.Update(p => p.Partitions[partition] = p.Partitions[partition] with
            {
                ScyllaCount = w, MaxTs = m,
                LastEntityId = lastRow.EntityId, LastKey = lastRawKey, LastTs = lastRow.Ts
            });
            _tracker.Update(p => p.MigratedRows += batch.Count);
        }
        _tracker.Update(p => p.Partitions[partition] = p.Partitions[partition] with { State = "migrated" });
    }

    private async IAsyncEnumerable<List<TsRow>> ReadBatchesAsync(
        string partition,
        long deltaFromTs,
        (Guid EntityId, string KeyRaw, long Ts)? resumeCursor,
        Dictionary<int, string> keyMap,
        bool hybridMode,
        [System.Runtime.CompilerServices.EnumeratorCancellation] CancellationToken ct)
    {
        if (deltaFromTs != long.MinValue)
            await foreach (var b in _reader.StreamPartitionAsync(partition, deltaFromTs, keyMap, hybridMode, _cfg.Migrator.PartitionBatch, ct))
                yield return b;
        else
            await foreach (var b in _reader.StreamPartitionPkAsync(partition, resumeCursor, keyMap, hybridMode, _cfg.Migrator.PartitionBatch, ct))
                yield return b;
    }

    // -------------------------------------------------------------------------
    // Phase 0 — load maps
    // -------------------------------------------------------------------------
    private async Task<(Dictionary<string,string> entityMap,
                         Dictionary<int,string>    keyMap,
                         bool                      hybridMode)> Phase0Async(CancellationToken ct)
    {
        Console.Error.WriteLine("[INFO] Phase 0: loading entity map and key map");
        var entityMap            = await _reader.LoadEntityMapAsync(ct);
        var (keyMap, hybridMode) = await _reader.LoadKeyMapAsync(ct);
        var total                = await _reader.CountTsKvAsync(ct);
        Console.Error.WriteLine(
            $"[INFO] Phase 0 complete: {entityMap.Count} entities, {keyMap.Count} keys, {total:N0} ts_kv rows");

        _tracker.Update(p =>
        {
            p.Phase        = "phase1";
            p.Phase1StartTs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        });
        return (entityMap, keyMap, hybridMode);
    }

    // -------------------------------------------------------------------------
    // Phase 1 — historical migration with 2-level parallelism:
    //   Level 1: N entities concurrently
    //   Level 2: N key-chunks per entity concurrently
    // -------------------------------------------------------------------------
    private async Task Phase1Async(
        Dictionary<string, string> entityMap,
        Dictionary<int, string>    keyMap,
        bool                       hybridMode,
        CancellationToken          ct)
    {
        int workers = _cfg.Migrator.Workers;
        Console.Error.WriteLine($"[INFO] Phase 1: historical migration started (workers={workers})");

        var completed = _tracker.Progress.CompletedEntities.ToHashSet();
        var allEntities = await _reader.GetDistinctEntitiesAsync(ct);
        var pending = allEntities.Where(e => !completed.Contains(e)).ToList();
        Console.Error.WriteLine(
            $"[INFO] Phase 1: {allEntities.Count} entities total, {pending.Count} pending");

        var outerOpts = new ParallelOptions { MaxDegreeOfParallelism = workers, CancellationToken = ct };

        await Parallel.ForEachAsync(pending, outerOpts, async (entityId, ct2) =>
        {
            if (_stop) return;

            if (!entityMap.TryGetValue(entityId, out var entityType))
            {
                Console.Error.WriteLine($"[WARN] entity {entityId} not in entity_map, skipping");
                _tracker.Update(p => p.SkippedRows++);
                return;
            }

            await ProcessEntityAsync(entityId, entityType, keyMap, hybridMode, workers, ct2);

            _tracker.Update(p =>
            {
                p.LastEntityId = entityId;
                p.CompletedEntities.Add(entityId);
            });
        });

        // ts_kv_latest (single pass — table is small)
        Console.Error.WriteLine("[INFO] Phase 1: migrating ts_kv_latest");
        var latestByType = new Dictionary<string, List<TsRow>>();
        await foreach (var row in _reader.StreamTsKvLatestAsync(keyMap, hybridMode, ct))
        {
            if (!entityMap.TryGetValue(row.EntityId, out var et)) continue;
            if (!latestByType.ContainsKey(et)) latestByType[et] = [];
            latestByType[et].Add(row);
            if (latestByType[et].Count >= _cfg.Migrator.BatchSize)
            {
                await _scylla.WriteLatestBatchAsync(latestByType[et], et, _cfg.Migrator.CastStrings, ct);
                latestByType[et] = [];
            }
        }
        foreach (var (et, rows) in latestByType)
            if (rows.Count > 0)
                await _scylla.WriteLatestBatchAsync(rows, et, _cfg.Migrator.CastStrings, ct);

        Console.Error.WriteLine("[INFO] Phase 1 complete");
        _tracker.Update(p => p.Phase = "live_sync");
    }

    private async Task ProcessEntityAsync(
        string                  entityId,
        string                  entityType,
        Dictionary<int, string> keyMap,
        bool                    hybridMode,
        int                     workers,
        CancellationToken       ct)
    {
        // Get this entity's distinct keys from ts_kv_latest (fast index lookup)
        await using var keyConn = new NpgsqlConnection(_connStr);
        await keyConn.OpenAsync(ct);
        var keyReader = new PgReader(keyConn, hybridMode);
        var keys = await keyReader.GetEntityKeysAsync(entityId, ct);

        if (keys.Length == 0) return;

        // Split keys into worker-sized chunks for intra-entity parallelism
        int chunkSize = Math.Max(1, (int)Math.Ceiling((double)keys.Length / workers));
        var chunks    = keys.Chunk(chunkSize);

        var innerOpts = new ParallelOptions { MaxDegreeOfParallelism = workers, CancellationToken = ct };

        await Parallel.ForEachAsync(chunks, innerOpts, async (keyChunk, ct2) =>
        {
            if (_stop) return;
            await using var conn = new NpgsqlConnection(_connStr);
            await conn.OpenAsync(ct2);
            var chunkReader = new PgReader(conn, hybridMode);

            foreach (var key in keyChunk)
            {
                if (_stop) return;
                await foreach (var batch in chunkReader.StreamEntityKeyAsync(
                                   entityId, key, keyMap, hybridMode, _cfg.Migrator.BatchSize, ct2))
                {
                    var partitions = await _scylla.WriteTsBatchAsync(
                        batch, entityType, _cfg.Migrator.Partitioning, _cfg.Migrator.CastStrings, ct2);
                    await _scylla.WritePartitionsAsync(partitions, ct2);
                    _tracker.Update(p => p.MigratedRows += batch.Count);
                }
            }
        });
    }

    // -------------------------------------------------------------------------
    // Phase 2 — live sync
    // -------------------------------------------------------------------------
    private async Task Phase2Async(
        Dictionary<string, string> entityMap,
        Dictionary<int, string>    keyMap,
        bool                       hybridMode,
        CancellationToken          ct)
    {
        Console.Error.WriteLine("[INFO] Phase 2: live sync started");
        long watermark = _tracker.Progress.WatermarkTs != 0
            ? _tracker.Progress.WatermarkTs
            : _tracker.Progress.Phase1StartTs - 60_000;
        _tracker.Update(p => p.WatermarkTs = watermark);

        while (!_stop && !ct.IsCancellationRequested)
        {
            long nowMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var latestByType = new Dictionary<string, List<TsRow>>();
            var tsByType     = new Dictionary<string, List<TsRow>>();

            await foreach (var row in _reader.StreamTsKvByTsAsync(watermark, keyMap, hybridMode,
                               _cfg.Migrator.BatchSize, ct))
            {
                if (!entityMap.TryGetValue(row.EntityId, out var et)) continue;
                if (!tsByType.ContainsKey(et))     tsByType[et]     = [];
                if (!latestByType.ContainsKey(et)) latestByType[et] = [];
                tsByType[et].Add(row);
                latestByType[et].Add(row);
                if (row.Ts > watermark) watermark = row.Ts;
            }

            foreach (var (et, rows) in tsByType)
            {
                var partitions = await _scylla.WriteTsBatchAsync(
                    rows, et, _cfg.Migrator.Partitioning, _cfg.Migrator.CastStrings, ct);
                await _scylla.WritePartitionsAsync(partitions, ct);
            }
            foreach (var (et, rows) in latestByType)
                await _scylla.WriteLatestBatchAsync(rows, et, _cfg.Migrator.CastStrings, ct);

            long lag = nowMs - watermark;
            _tracker.Update(p => p.WatermarkTs = watermark);

            if (lag < _cfg.Migrator.LagThresholdMs)
                Console.Error.WriteLine($"[INFO] LAG {lag}ms < {_cfg.Migrator.LagThresholdMs}ms — ready for switchover");

            await Task.Delay(TimeSpan.FromSeconds(_cfg.Migrator.LiveSyncInterval), ct);
        }
    }

    public void RequestStop() => _stop = true;
}
